### ISSUE TITLE
Clean up Client and Server integration tests

### ISSUE THREAD

**Issue by seanmonstar** [hyperium/hyper#1675]
Both the client and server tests (`tests/client.rs` and `tests/server.rs`) started with small abstractions, but they weren't flexible enough, and so as more complicated cases were needed, the tests just got more and more complicated, with a lot of repetition. It'd be huge to find a way to refactor them.

**Comment by JeffLabonte** [comment]
I will take this issue and work on it!

**Comment by tiendq** [comment]
@seanmonstar is it still open? :)

**Comment by seanmonstar** [comment]
Yes, this is still open because the integration tests feel very messy, and there's a lot of repeated code. I'm sorry I haven't been able to provide much guidance on how to improve them, if I thought about it a lot, I'd probably fix them myself :)

**Comment by tiendq** [comment]
Well, if it's very messy so we could improve it a little each PR, for example let's remove (as many as possible) repeated code first. If you feel ok with this approach I could try it :) (hopefully it's easy as labelled :D).

**Comment by ryanyz10** [comment]
I'd like to try and tackle this one if it hasn't been done yet! The tests seem quite long though...

**Comment by ryanyz10** [comment]
Ok great, I'll slowly work on this then.

**Comment by alexs-sh** [comment]
Hi @seanmonstar
Could you please take a look at https://github.com/hyperium/hyper/pull/3072

This PR doesn't provide any significant changes but allows to make one step in decreasing code duplication. If it's the wrong way, please, just reject the PR. If it brings something useful, then I can continue with it.

Thanks
